add metrics to OperationPool (need to decline this )

seroze opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/ConsenSys/teku/blob/master/CONTRIBUTING.md -->

## PR Description
Fix for issue 3917 (Metrics for operation pool size and usage)
## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->
Metrics for operation pool size and usage #3917
## Documentation

- [ ] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

rolfyone committed:
rename syncCommitteeIndices in sync committee duties in line with api change (#3934)

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
implement POST `/eth/v1/validator/contribution_and_proofs` (#3933)

* implement POST `/eth/v1/validator/contribution_and_proofs`

does not have the validator api handler implemented, will throw a not supported exception if called.

Partially addresses #3928

Signed-off-by: Paul Harris <paul.harris@consensys.net>

mbaxter committed:
[Issue-3905] Implement version2 of GetMetadata RPC (#3936)

ajsutton committed:
Implement sendSignedContributionAndProofs (#3938)

ajsutton committed:
Consider nodes in sync if they are close to chain head (#3940)

rolfyone committed:
implement GET `/eth/v1/validator/sync_committee_contributions` (#3939)

* implement GET `/eth/v1/validator/sync_committee_contributions`

partially addresses #3929

Signed-off-by: Paul Harris <paul.harris@consensys.net>

mbaxter committed:
[Issue-3806] Fix uint to bytes conversion bug (#3941)

ajsutton committed:
Perform Sync Committee Aggregation Duties (#3935)

rolfyone committed:
Update v2 getBlock to return a correctly formatted altair block. (#3947)

Also raise 400 error if v1 getBlock is used to fetch an altair block.

Started obfuscating schema object creation into the provider layer so that the spec can be used to make some decisions.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Use the correct slot when creating sync aggregates (#3948)

ajsutton committed:
Improve Altair validator logging output. (#3949)

Shorten sync committee type names and update the longest type in ValidatorLogger.

ajsutton committed:
Update validator client to switch Fork instances during forks (#3952)

ajsutton committed:
Ensure the state used to calculate duties for an epoch is after the fork slot. (#3955)

ajsutton committed:
Convert to subnet id when processing sync committee subscriptions (#3957)

ajsutton committed:
Restrict duty requests to current and next sync committee period (#3956)

ajsutton committed:
Include epoch info in message when state is unsuitable for calculating committee duties. (#3958)

ajsutton committed:
Fallback to RocksDB if LevelDB is not supported on the platform. (#3963)

Our current LevelDB library does not include binaries for ARM systems whereas RocksDB does, so preserve compatibility by defaulting to RocksDB on ARM.

ajsutton committed:
Log a message when network upgrades activate. (#3962)

ajsutton committed:
Add cli option to override the altair fork block and enable it (#3961)

ajsutton committed:
Ensure the state used to calculate sync committee duties is not before the altair for slot. (#3964)

rolfyone committed:
create an unsigned block for altair (#3960)

* create an unsigned block for altair

fixes #3953

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Remove changelog items released as part of 21.5.0. (#3966)

rolfyone committed:
update rest postBlock to support altair blocks (#3965)

* update rest postBlock to support altair blocks

Signed-off-by: Paul Harris <paul.harris@consensys.net>

NicolasMassart committed:
update dependencies (#3968)

tbenr committed:
Adds error detection in web3j responses from eth1 endpoint (#3969)

ajsutton committed:
Improve error message when keystore password file not found (#3970)

rolfyone committed:
fix swagger documentation for block endpoints (#3972)

* fix swagger documentation for block endpoints

improved swagger so that there's less duplication of schema tags

Signed-off-by: Paul Harris <paul.harris@consensys.net>

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

ajsutton committed:
Retrieve the hashTreeRoot from the unmodifiable preState to fully benefit from caching. (#3973)

ajsutton committed:
Add additional pyrmont bootnodes (#3976)

ajsutton committed:
Offset sync committee duties by a slot (#3979)

rolfyone committed:
sync committee subscription fix (#3980)

epoch / slot translation would mean we were never subscribed to subnets basically, fixed a test so correct behavior is now demonstrated.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Fix validation of sync committee gossip that refers to a block root just before the Altair fork. (#3975)

rolfyone committed:
Update remote validator client to handle altair block creation (#3978)

* Update remote validator client to handle altair block creation

Added a hidden cli option to direct the validator client to use `eth/v2/validator/blocks/:slot`, so that the version field in the response can be used to determine which type of block should be deserialized.

The `eth/v1/validator/blocks/:slot` endpoint does return an altair block but its less simple to decode, which ultimately stopped the remote validator from participating in block production.

The block production v2 endpoint is part of the standard spec, but we need to be careful switching to it, so for now this option is hidden and disabled by default.

For deserialization to work, a custom deserializer has been added for the GetNewBlockResponseV2 object.

it appears theres still functionality broken for remote validators, but this is a good start.

fixes #3974

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
trigger onHeadUpdate in event source handler (#3983)

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Simplify how block production metrics are calculated (#3984)

Only need to store the slot and block root then compare to the historic block roots in the head state rather than loading the entire canonical block from disk.

rolfyone committed:
implement remote call to get sync committee contributions for validator client (#3985)

* implement remote call to get sync committee contributions for validator client

 - needed to change sync committee contributions to bytes so that it could be correctly treated for the round trip.

 - added a validator test for creating sync committee contribution and reading it into the validator client.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

NicolasMassart committed:
fix yaml load (#3987)

include config in try to catch issues and fail job if any
some code formatting

ajsutton committed:
Fix calculation of block performance metrics when head block was locally produced. (#3986)

ajsutton committed:
Remove reference to removed http module (#3991)

Run checkMavenCoordinateCollisions first to catch this problem in CI.

ajsutton committed:
Add toVersionAltair to BeaconBlockBody. (#3992)

mbaxter committed:
[Issue 3925] Search for peers on required sync committee subnets (#3989)

ajsutton committed:
Simplify BlockProcessor (#3997)

Reduce scope of public interface.
verify/validate methods consistently return BlockValidationResult rather than a mix of return values and throwing exceptions.

ajsutton committed:
Report sync committee performance (#3995)

rolfyone committed:
implement GET `/eth/v1/beacon/states/{state_id}/sync_committees` (#3993)

* implement GET `/eth/v1/beacon/states/{state_id}/sync_committees`

fixes #3990

Signed-off-by: Paul Harris <paul.harris@consensys.net>

mbaxter committed:
Cut deprecated logic from AttesterSlashingValidator (#3999)

NicolasMassart committed:
add link checker (#4004)

fixes #3831

ajsutton committed:
Expose a single getBlockProcessor method on spec instead of individually delegating each of its methods. (#4001)

Makes fuzzutil and reference test code a bit more verbose but simplifies the production code in Spec and makes it a lot easier to see where BlockProcessor methods are actually being used.

ajsutton committed:
Move computeStartSlotAtEpoch to MiscHelpers to match its location in the spec. (#4007)

mbaxter committed:
[Issue 3356] Cut deprecated AttestationUtil (#4006)

ajsutton committed:
Move get block root to BeaconStateAccessors (#4008)

ajsutton committed:
Rename SyncCommitteeSigningData to SyncAggregatorSelectionData. (#4009)

ajsutton committed:
Include SyncAggregate in block batch signature verification (#4010)

Replace boolean argument controlling whether to validate signatures and state roots with two separate methods and the ability to pass in a `BLSSignatureVerifier`.
The signature verifier is then also available through the entire block processing process, not just in the secondary verification stage so can be used when verifying sync committees which would otherwise need to perform duplicate work.

ajsutton committed:
Upgrade to v1.1.0-alpha.5 spec release (#4002)

See https://github.com/ethereum/eth2.0-specs/releases/tag/v1.1.0-alpha.4 for key changes (alpha.5 just fixed a couple of minor issues in the alpha.4 release)

Key changes:
* Support new transition reference tests
* Replace sync committee aggregates list with single aggregate public key
* Migrate validator participation records to Altair format during upgrade
* Update inactivity penalty deltas processing
* Fix sync committees for Altair genesis
* Reduce sync committee size and subnet count

rolfyone committed:
add `/eth/v1/node/peer_count` endpoint (#4015)

Due to the way connections are maintained at a different layer, it's likely that it'll be mostly 0's with 'connected' being maintained, but I followed the standard response structure.

fixes #4012

Signed-off-by: Paul Harris <paul.harris@consensys.net>

NicolasMassart committed:
Update GH Action link check to weekly schedule (#4016)

update cron to once a week on sunday

Commit:
Customize health syncing status via parameter (#4013)

- use request parameter to override default syncing status
- return default status if wrong parameter passed
- add tests to cover different scenarios

ajsutton committed:
Use default charset for console output (#4014)

Previously forced output to UTF-8 which may corrupt characters on systems using a non-ascii based charset.

rolfyone committed:
implement GET `/eth/v2/debug/beacon/states/:state_id` (#4000)

* implement GET `/eth/v2/debug/beacon/states/:state_id`

 - ensure `/eth/v1/debug/beacon/states/:state_id` doesnt return altair state as json.

 - `BeaconState` is now an abstract base class and `/eth/v1/debug/beacon/states/:state_id` returns `BeaconStatePhase0` instead

fixes #3990

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Fix NullPointerException when requesting genesis data prior to genesis (#4017)

rolfyone committed:
added slashing protection repair command (#4011)

* added slashing protection repair command

fixes #3868

Signed-off-by: Paul Harris <paul.harris@consensys.net>

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

ajsutton committed:
Rename classes shared by RocksDB and LevelDB (#4024)

ajsutton committed:
Fix paths used to determine split of reference tests across executors (#4028)

Adjust parallelism for referenceTest and acceptanceTest tasks.

ajsutton committed:
Remove createUnsignedAttestation API and create AttestationData directly (#4029)

mbaxter committed:
Use standard error when blocks are unavailable (#4027)

mbaxter committed:
[Issue-3356] Reorganize committee logic (#4026)

mbaxter committed:
[Refactor] Move utils to MiscHelpers (#4033)

ajsutton committed:
Remove pow dependency from spec (#4030)

Split out the API classes into :ethereum:pow:api. This will allow the actual pow chain logic to use the spec classes instead of the old style constants.

ajsutton committed:
Fix inactivity score calculation (#4021)

Support new reference test types coming in future releases and improve error messages.

benjaminion committed:
Remove mikuli (#4035)

* Remove Mikuli BLS implementation

* Update changelog

* Tidy up blst unavailable handling.

* Fix spotless.

Co-authored-by: Adrian Sutton <adrian.sutton@consensys.net>

ajsutton committed:
Fix handling of eth_getLogs errors that return too many logs (#4041)

rolfyone committed:
Allow a clock disparity for sync committee signatures (#4043)

fixes #4039

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Add metric to report number of eth1 requests sent (#4044)

ajsutton committed:
Add time based eth1 head tracker (#4040)

Currently toggled off.

mbaxter committed:
[Issue 3356] Deprecated BeaconStateUtil cleanup (#4037)

mbaxter committed:
[Issue 4023] Rework config file format (#4042)

rolfyone committed:
Altair acceptance test (#4045)

* Altair acceptance test

 - ensure sync committee aggregates are set on an altair network

 - demonstrate remote beacon nodes can connect, and remote validators are functional on altair

fixes #4018

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Add support for retrieving blocks as ssz to v2 api (#4046)

ajsutton committed:
Fix copying of current justified checkpoint during Altair transition (#4047)

rolfyone committed:
Subscribe to sync committee subnets (#4049)

* Subscribe to sync committee subnets
 - if --p2p-subscribe-all-subnets-enabled is set, subscribe to all sync committee subnets
 - increase AT timeout to help reduce intermittent failures.

fixes #4036

Signed-off-by: Paul Harris <paul.harris@consensys.net>

usmansaleem committed:
Upgrade gradle and various plugins (#4038)

- Gradle build enhancements. 
- Upgraded gradle and various plugin versions. 
- Introduced new license dependency report generator with custom renderer. 
- Refactored custom errorprone checks to its own repo. 
- The project can now be built against JDK 16.

rolfyone committed:
add syncnets to network identity metadata for altair (#4053)

fixes #4052

Signed-off-by: Paul Harris <paul.harris@consensys.net>

mbaxter committed:
[Issue 4023] Migrate new domain constants (#4050)

ajsutton committed:
Upgrade reference tests to alpha.6 (#4051)

ajsutton committed:
Add CONFIG_NAME back into specs returned from REST API (#4054)

mbaxter committed:
[Issue 4023] Migrate phase0 validator constants (#4057)

mbaxter committed:
[Issue 4023] Migrate altair validator constants (#4058)

mbaxter committed:
[Issue 3356] Update operation signature verifiers (#4059)

mbaxter committed:
[Issue 3356] Update operation validators (#4060)

rolfyone committed:
Add sync committees to the list of filtered topics for gossip (#4064)

* Add sync committees to the list of filtered topics for gossip

 - also ensure remote validators are requesting sync committee subscriptions

 - fix acceptance test to use correct number of validators

partially addresses #4055

Signed-off-by: Paul Harris <paul.harris@consensys.net>

mbaxter committed:
[Issue 4034] Fix threading issue in LevelDbTransaction (#4065)

mbaxter committed:
[Issue 3356] Limit usages of deprecated ValidatorsUtil (#4066)

rolfyone committed:
cleanup the exceptions at the end of acceptance test (#4069)

* cleanup the exceptions at the end of acceptance test

Signed-off-by: Paul Harris <paul.harris@consensys.net>

mbaxter committed:
[Issue 3356] Delete deprecated CommitteeUtil (#4068)

rolfyone committed:
update to alpha.7 altair spec (#4070)

* update to alpha.7 altair spec

fixes #4063

ajsutton committed:
Remove the now empty data module. (#4072)

It's just a folder grouping other modules now.

ajsutton committed:
Remove released changes. (#4073)

mbaxter committed:
[Refactor] Rework topic management (#4071)

rolfyone committed:
Eth1 min genesis block failure should not be fatal (#4074)

* Eth1 genesis block failure should not be fatal

- when looking for the eth1 genesis block, if the eth1 node is not responding with blocks, try again after delay but log a message ot indicate the eth1 node may not be sync'd.

fixes #3924

Signed-off-by: Paul Harris <paul.harris@consensys.net>

benjaminion committed:
Cache public key group membership status (#4075)

mbaxter committed:
[Refactor] Make GossipTopicName an enum (#4076)

ajsutton committed:
Update to discovery 0.4.7 (#4082)

Includes fixes for publishing the correct ENR sequence number and update peer's ENRs when their sequence number increases.

rolfyone committed:
Change `/eth/v1/node/health` to notify 206 during startup (#4083)

* Change `/eth/v1/node/health` to notify 206 during startup

 - During startup before a node has peers, node/health was notifying it was healthy, even though the node wasn't able to service requests.

 - similarly we would potentially have been able to accept a block to post or provide attester and proposer duties when we shouldn't really be performing these actions during startup

 fixes #4078.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Ensure that fork choice has run prior to creating attestations (#4084)

This only makes a difference when using dependent roots as otherwise fork choice is automatically run before sending the event that triggers attestation creation. Not running fork choice prior to creating attestations results in a reduction in attestation rewards which this change fixes.

vishaljhala committed:
Ignore launchpad deposit data json file (#4085)

Co-authored-by: Adrian Sutton <adrian.sutton@consensys.net>

ajsutton committed:
Update node dependencies. (#4086)

mbaxter committed:
[Issue 4067] Update gossip message ids (#4081)

ajsutton committed:
Add metrics for eth1 voting (#4087)

ajsutton committed:
Revert "Update node dependencies. (#4086)" (#4092)

This reverts commit 584a25b7a5bf40ceb9388f5ee00024245d0ee7c3.

mbaxter committed:
[Issue-4097] Consider subnet id when processing syncCommittee gossip (#4088)

Co-authored-by: Adrian Sutton <adrian.sutton@consensys.net>

Commit:
Add readiness endpoint (#4093)

- add new readiness endpoint

rolfyone committed:
Add raw handling to database iterators and accessors (#4096)

* Add raw handling to database iterators and accessors

This will enable more efficient copying between database handles, which is directly related to #3882, to allow a database migration path.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

mbaxter committed:
[Issue-3356] Deprecated BeaconStateUtil cleanup (2) (#4100)

ajsutton committed:
Enable time based eth1 follow distance (#4098)

benjaminion committed:
Avoid zero batch multiplier (#4114)

vishaljhala committed:
Validate initial state matches fork schedule (#4109)

mbaxter committed:
[Issue-4101] Batch verify attestation gossip (#4106)

mbaxter committed:
Use consistent thread count (#4115)

ajsutton committed:
Deafult batch verify signatures to on (#4116)

rolfyone committed:
CLI to allow conversion between database versions (#4110)

* CLI to allow conversion between database versions

Implemented migrate-data subcommand to create a new database version.

The `--network` context is needed. It's probably easiest to specify `--config-file` as part of the CLI like normal, then execute the migrate-data subcommand, so the path, eth1 contract etc can be read from the config file while the new database is being created.

 - cannot create a database instance of the same type currently.

 - prints status updates every 100k records, and also a count of rows moved for each column.

 - no data validation is performed, byte data is literally copied. This makes the move fairly fast.

 - if an existing beacon.old is found, the program will abort. Beacon.old is theoretically a real database that was working, so removing it might be dangerous.

 - if there's an existing beacon.new, it's cleaned up at the start of the process - this was likely an incomplete conversion.

 - if the subcommand completes successfully, the 'beacon' folder will be the new database type. 'beacon.old' will be the original database, and will need to be manually cleaned up by the user.  It shouldn't be cleaned up until the new database is verified to be ok.

 - the 'beacon' folder is impacted by this subcommand, but other folders (eg. validator are not touched).

fixes #3882

Signed-off-by: Paul Harris <paul.harris@consensys.net>

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

ajsutton committed:
Remove released items from changelog. (#4120)

rolfyone committed:
Altair spec sync committee changes (#4117)


Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Enable dependent roots by default (#4121)

mbaxter committed:
[Issue-3356] Clean up deprecated BeaconStateUtil (3) (#4102)

mbaxter committed:
Cut unused code (#4123)

mbaxter committed:
Split batches recursively on batch verification failure (#4124)

ajsutton committed:
Remove EventBus (#4125)

Main change is to switch BlockImportedEvent over to event channels via BlockImportNotifications.

usmansaleem committed:
External signer Altair support (#4107)

* Send metadata information to external signer that can be used to recalculate signing root
* External signer integration test for Altair support

ajsutton committed:
Remove compatibility tests. (#4128)

We haven't run these for quite some time and can test against other clients much more easily elsewhere.

ajsutton committed:
Remove more code from utils module (#4127)

ajsutton committed:
Enable Altair if it is present in the spec config (#4132)

CLI override option is no longer required if Altair is enabled in the config.

ajsutton committed:
Periodically ping bootnodes (#4131)

Ensures we recover after a network outage that causes all discovery nodes to be dropped.

ajsutton committed:
Upgrade reference tests to alpha 8. (#4130)

ajsutton committed:
Use current chain head when validating sync committee gossip (#4142)

ajsutton committed:
Use batch signature verification for sync committee signatures (#4144)

ajsutton committed:
Upgrade jvm-libp2p to 0.8.3 - fixes file handle leak (#4145)

ajsutton committed:
Use SecureRandom as part of BLS batch verification. (#4143)

ajsutton committed:
Log incorrect deposit sequence errors to console (#4146)

ajsutton committed:
Fix PostSyncCommittees to return 200 when there are no errors instead of 503 (#4133)

rolfyone committed:
Serialize epoch participation as string list (#4147)

* Serialize epoch participation as string list

fixes #4095

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Avoid declaring referenceTest/resources directory twice (#4149)

ajsutton committed:
Remove caching for meta data messages. (#4151)

There are now two forms of metadata message so caching is more complex and given the messages are cheap to create and not created very often the cache doesn't provide enough value.

ajsutton committed:
Use Eth1HeadTracker to find initial chain head (#4148)

Previously reimplemented follow distance using the block number strategy.

ajsutton committed:
Migrate usages of Constants to SpecConfig (#4129)

rolfyone committed:
add sync_committee_subnet_count to spec/config output (#4152)

fixes #4141

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Add MIN_ANCHOR_POW_BLOCK_DIFFICULTY to the list of config items we ignore. (#4156)

ajsutton committed:
Add metric to report the number of correct signatures produced in an epoch (#4154)

ajsutton committed:
Add metric to report the number of signatures included in the head block sync aggregate (#4153)

rolfyone committed:
return error when contributions are invalid (#4155)

* return error when contributions are invalid

Previously when a contribution and proof was rejected, the error reason was only in trace. Now the errors should be returned, in a semi-colon separated list if there are multiple.

Ignored entries will still not result in an error being returned to the user.

fixes #4138

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Upgrade reference tests to beta.1 (#4157)

ajsutton committed:
Optimise collection of Altair validator stats (#4158)

rolfyone committed:
Add checkarg for empty public keys being added to batch verify (#4160)

partially addresses #4105

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Ignore MIN_ANCHOR_POW_BLOCK_DIFFICULTY in ConstantsReader (#4159)

ajsutton committed:
Include information about the rejected block when state transition fails (#4163)

protolambda committed:
Fix null gossipType in Eth2TopicHandler (Altair) (#4165)

rolfyone committed:
Dont call signature verifier without public keys (#4169)

Dont call signature verifier without public keys

When verifying sync aggregates, avoid calling through to signature verification if the pubkeys list is empty, as it throws an exception.

Added unit test cases to show desired behavior, and a comment explaining why eth2_fast_aggregate_verify is slightly different to the altair spec.

fixes #4105

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Add previous_live_sync_committee metric (#4161)

ajsutton committed:
Increase target sync subcommittee aggregators to 16 (#4162)

ajsutton committed:
Reject sync committee contributions with no participants (#4171)

Previously threw an exception because BLS.verify doesn't handle empty signatures.

ajsutton committed:
Rename SyncCommitteeSignature to SyncCommitteeMessage (#4170)

ajsutton committed:
Add option to log warnings when P2P messages are invalid/rejected (#4172)

ajsutton committed:
Add warning with full details for invalid blocks. (#4174)

ajsutton committed:
Skip logging warnings when blocks are from the future (#4177)

ajsutton committed:
Provide meaningful rejection reasons (#4179)

rolfyone committed:
Relocate signature verification for batch storage (#4183)

* Relocate signature verification for batch storage

Signature verification was inside single threaded storage, and slowing down storage.

Now located in the threaded batch fetcher.

fixes #4019

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Upgrade discovery to 0.4.8 (#4185)

* Include found peers in the consideration of which peers to connect to.
* Known nodes are only stored in KBuckets, ensuring they don't grow unbounded.
* Implement recursive search correctly.

ajsutton committed:
Upgrade reference tests to beta 2 (#4188)

ajsutton committed:
Use thread safe datastructures for performance tracking (#4186)

rolfyone committed:
cleanup changelog after 21.7.0 release (#4194)

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
add contribution_and_proof events in the rest api. (#4189)

* add contribution_and_proof events in the rest api.

When in Altair, contribution_and_proof events are now able to be distributed in the rest api events.

fixes #4178

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Update docker images (#4195)

* Adds jdk16 docker image
* Adds curl to all docker images

rolfyone committed:
Empty body should return bad request (#4197)

* Empty body should return bad request

When an empty body is received for an endpoint requiring a body of data, a bad request should be the result, rather than a 500 error.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

rolfyone committed:
use event topics to show contribution gossip (#4198)

* use event topics to show contribution gossip

Now that there is a contribution_and_proof http event, listen for it during an acceptance test to show that contribution_and_proof objects are coming from validators remote to the beacon node, indicating that the contributions are being successfully gossipped.

fixes #4191

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Fix potential race condition in DefaultPerformanceTracker (#4201)

start is called from a different thread to onSlot so need to ensure a consistent view of nodeStartEpoch.

rolfyone committed:
Add development option to disable early attestation production. (#4205)

* Add development option to disable early attestation production.

By default attestation production occurs as soon as a block is produced, but with dependent roots, it is sometimes useful to be able to tweak this default behaviour.

Added internal CLI option `--Xvalidators-early-attestations-enabled` which defaults to true.

fixes #4193

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Fix InvalidDepositEventsException after restart (#4206)

Handles the case where the current canonical head is prior to the last deposit replayed from storage. Previously we're re-request deposits from the last deposit back to the canonical head.

ajsutton committed:
Remove genesis, far future and domain constants from Constants (#4207)

ajsutton committed:
Only advertise address in ENR when external address is known (#4212)

ajsutton committed:
Use Spec in BlockValidator (#4208)

ajsutton committed:
Switch to AdoptOpenJDK, Ubuntu docker base images (#4215)

ajsutton committed:
Add Altair fork epoch to Pyrmont config (#4210)

Switches Pyrmont config to use the newer preset style.

rolfyone committed:
Move previously experimental endpoints to their correct locations (#4211)

Also deprecated old endpoints in favour of the new endpoints, as the old ones will cease to function when altair comes online.

fixes #3703

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Remove unused validatorIndex param for sendSignedAttestation (#4218)

ajsutton committed:
Update dependencies (#4216)

Updates dependencies that have minor CVEs raised against them to versions containing fixes. Its unlikely that any of the CVEs would apply to Teku's usage but worth tidying up.

ajsutton committed:
Fix handling of sendSyncCommitteeMessage responses. (#4220)

Ok responses are expected to be empty and bad request responses are expected to have the failure response content.

rolfyone committed:
Make url max request length configurable (#4222)

* Make url max request length configurable

By default, embedded jetty allows 8192 (8K) bytes for a request length.

Increased the default request length for teku to 65535 (64K) bytes, and added a development CLI arg to allow it to be configured.

'--Xrest-api-max-url-length': (min: 4096, max: 1052672) - hopefully these will provide sane guidelines for users that need to tweak the url length.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

regio committed:
Updating mockito to 3.11.2 (#4224)

rolfyone committed:
If atomic move fails, perform a normal move (#4221)

* Move previously experimental endpoints to their correct locations

Also deprecated old endpoints in favour of the new endpoints, as the old ones will cease to function when altair comes online.

fixes #3703

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
cleanup changelog after 21.8.0 (#4231)

rolfyone committed:
Make validators-early-attestations-enabled a public option (#4233)

* Make validators-early-attestations-enabled a public option

fixes #4225

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton committed:
Add additional bootnodes for Prater (#4235)

ajsutton committed:
Upgrade discovery (#4236)

Includes improvement to consider all authenticated sessions as potential connection targets instead of only those in the node table.
Add reason to peer disconnect logs.

rolfyone committed:
clean up migrate-database command line arguments (#4234)

* clean up migrate-database command line arguments

It was difficult to determine the command line arguments actually required for the migrate-database command, and several listed options were completely ignored.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Add progress counter to some of the migrate-database columns (#4238)

* Add progress counter to some of the migrate-database columns

Particularly finalized state and block columns can be time consuming to copy, so providing some context of the number of elements and percentage completed is helpful to understand how long is likely to be remaining in the process.

fixes #4229

Signed-off-by: Paul Harris <paul.harris@consensys.net>

regio committed:
Updating syncing message (#4237)

* Updating syncing message

Fixes #4228

* Updating changelog.

Co-authored-by: Adrian Sutton <adrian.sutton@consensys.net>

ajsutton committed:
Remove invalid sync contribution validation. (#4239)

SignedContributionAndProofValidator was incorrectly rejecting contribution gossip if the chain head state slot was greater than the contribution slot. This check could fail because of a race condition when contributions were received right at the end of the slot.

ajsutton committed:
Report validation failures for submitted attestations (#4219)

ajsutton committed:
Support voting for eth1 blocks prior to the first deposit. (#4245)

On testnets, it's common for a genesis state to be generated that already includes validators, so the chain can start without having any deposits received.

ajsutton committed:
Handle exception when peers send additional responses cleanly (#4246)

Previously an IllegalStateException was thrown which is unhandled. Now a RpcException is thrown to indicate that the remote peer has behaved incorrectly. The request is aborted and logged is at debug level.

seroze committed:
add metrics to OperationPool

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java:
```diff
@@ -31,16 +35,32 @@
+        metricsSystem.createLabelledCounter(
+            TekuMetricCategory.BEACON,
+            OPERATION_POOL_SIZE_VALIDATION_REASON + metricType,
+            "track validation reason");
```
This may be causing issues - `track validation reason` won't be a valid metric label name because of the spaces.  I'd suggest just `result`.

ajsutton commented on ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java:
```diff
@@ -31,16 +35,32 @@
   private static final int OPERATION_POOL_SIZE = 1000;
+  private static final String OPERATION_POOL_SIZE_METRIC = "eth1_operation_pool_count_for_";
+  private static final String OPERATION_POOL_SIZE_VALIDATION_REASON =
+      "eth1_operation_pool_validation_reason_for_";
```
I'd probably name these `operation_pool_size_` and `operation_pool_validation_`

seroze committed:
Add message when syncing operation has completed. (#4247)

* Add message when syncing operation has completed.

seroze committed:
Add message when syncing operation has started. (#4255)

* Add message when syncing operation has started.

seroze committed:
Base docker images on ubuntu:latest (#4250)

* All docker images now install a JRE and Teku on top of `ubuntu:latest` to ensure we have the latest OS security updates.
* jdk16 docker image now uses a JDK eclipse-tumerin.
* jdk15 and jdk14 images are back to including OpenJDK builds
* Added note to changelog that jdk14 and jdk15 images will be removed in the future.

seroze committed:
Use IntList instead of List<Integer> (#4259)

Avoids creating a lot of Integer wrapper objects.

seroze committed:
scan docker images for vulnerabilities (#4253)

* scan docker images for vulnerabilities

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Remove changelog items from 21.8.1. (#4249)

seroze committed:
Reduce complexity and object creation in VoteTracker (#4254)

There are a lot of VoteTracker instances created and stored, to improve performance we already use a custom serializer when storing to the database so don't need them to use the SSZ framework.  Making them a POJO simplifies the code, but also avoids the need for boxing of the Bytes32 and UInt64 values they contain so reduces GC load as they are created and destroyed.

seroze committed:
Optimise processEffectiveBalanceUpdates (#4256)

Identify when a validator is already at the maximum effective balance and avoid setting to the same value. Avoids creating new SSZ views and invalidating caches unnecessarily.

seroze committed:
Optimise voted validator iteration (#4260)

Store the highest index of any validator that has attested to avoid copying the entire set of indices. Since nearly every validator has voted at some point and we can quickly identify validators that haven't voted there's no benefit to iterating over the specific list, and copying it for thread safety creates a significant amount of garbage.

seroze committed:
remove docker scan to avoid spurious error messages (#4261)

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
add dockerScan task to circleCI using trivy (#4267)

* add dockerScan task to circleCI using trivy

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Precompute SszLengthBounds (#4262)

seroze committed:
Validator not found should return a 404 from rest api (#4265)

* Validator not found should return a 404 from rest api

fixes #4263

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Removed deprecated constants from sync package. (#4268)

* Removed deprecated constants from sync package.

seroze committed:
Fixed building instructions (#4270)

Long time ago there was a script to clone the repo, but it's gone away with #1193

seroze committed:
Add feature toggle to send attestations as a batch (#4252)

seroze committed:
Add slack notification to nightly build on failure (#4274)

Also added medium severity so that it'll fail next run, and tweaked the time to run.

Will change back the time to run once we've seen the failure and can remove 'medium'.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Add notify task to master build failure (#4277)

* add notify task to master build failure

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Requesting an attestation for a future slot should not result in a 500 error (#4276)

* Requesting an attestation for a future slot should not result in a 500 error

When creating a new attestation, any illegal argument exceptions should resolve to a 400 error.

fixes #4269

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
add bash to the image running trivy during nightly task (#4279)

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
fix name of data-storage-non-canonical-blocks-enabled command line option (#4284)

fixes #4283

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Progress removal of constants from the Constants object. (#4280)

Moved BLS_WITHDRAWAL_PREFIX and DEPOSIT_CONTRACT_TREE_DEPTH to NetworkConstants.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
External signer Altair block (#4282)

* External signer Altair support

 -- send milestone and Altair specific block to external signer

Signed-off-by: Usman Saleem <usman@usmans.info>

seroze committed:
Ensure all CLI options display default values in help output (#4272)

Also improves the help text for some command options.

seroze committed:
Add Altair fork epoch for Prater. (#4285)

Upgrades Prater to new preset based config.

seroze committed:
Introduce SendingStrategy for aggregates (#4278)

seroze committed:
Verify SSZ bit is padded to byte boundary with zeros (#4287)

SSZ bits, when not packed as part of a list or vector, should be either 0x00 or 0x01 but Teku previously wasn't verifying the leading bits padding out the byte were all zero.

seroze committed:
Update docker creation to not use layer caching (#4297)

* Update docker creation to not use layer caching

Also change the nightly branch back to master, as gh-pages uses a different CI.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Nightly branch change (#4294)

* run nightly on a different branch

This will stop it reporting that master has failed. The images checked are remote and generated from master, but its unhelpful to show our build broken when the nightly fails.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Add new BLS reference tests (#4248)

Co-authored-by: Ben Edgington <ben@benjaminion.xyz>

seroze committed:
Report validation errors from POST aggregates API (#4281)

Also use batch sending strategy for aggregates when feature toggle is on.

seroze committed:
Remove released items from changelog (#4292)

seroze committed:
introduce metricType in OperationPool constructor to handle Collector already Registered Failure

seroze committed:
pull docker image before scanning (#4301)

* docker pull before trivy scan, so that we are working on the latest build

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Remove support for V6/leveldb2 databases with separate archive databases (#4290)

seroze committed:
Use atomic write for signing records (#4293)

* Use atomic write for signing records

seroze committed:
Additional validation when applying attestations from blocks (#4300)

Avoids applying attestations received via blocks if they are not from the current or previous epoch and ensures the full validate_on_attesation rules are checked.

seroze committed:
Update to beta.3 reference tests. (#4298)

Still ignoring some new test formats which need new test executors implemented.

seroze committed:
Check slot number when calling getStateSyncCommittees (#4302)

Also check slot number is an altair slot on GetSyncCommitteeContribution.

Provides a fast response when non altair slots are passed for these endpoints, rather than potentially having to wait for the state to be generated.

fixes #4296

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
fix gradlew spotlessApply

seroze committed:
Justified block is always a viable head (#4304)

seroze committed:
Add test executors for new BLS reference tests (#4303)

Upgrades BLST to latest version.

seroze committed:
set timeOut to 6 minutes

seroze committed:
refactor metric names and remove spaces in help field

seroze committed:
increase timeout to 3 minutes as shouldChangeValidatorStatusAfterSubmittingVoluntaryExit() is failing due to timeout

seroze committed:
increase timeOut to 10mintues

seroze committed:
Make use of abstract unit tests for BLST implementation (#4242)

* Make use of abstract unit tests for BLST implementation
* Return back aggregateVerifyDuplicateMessages() test and prohibit duplicate messages in aggregateVerify() for BLST impl
* Remove duplicate message checks and corresponding test as we are under ProofOfPossession BLS scheme
Co-authored-by: Ben Edgington <ben@benjaminion.xyz>

seroze committed:
Optimise validator index checking (#4306)

* Optimise validator index checking

Also increase batch size to 50 validator keys from 10, to reduce remote calls.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

seroze committed:
Wiring for trie based state database format (#4289)

Makes it possible to plug in different finalized state storage approaches to the existing finalized dao.

seroze committed:
fix spotlessApply

seroze committed:
merge with master
